packages/sdk/app-framework/src/plugins/plugin-intent/intent.ts

```diff
@@ -2,100 +2,134 @@
-   */
-  // TODO(burdon): Chainable types? (see Effect hooks).
-  result?: any;
-};
+// TODO(wittjosiah): Could input/output types be allowed to be primitives and not strictly structs?
+//   This would allow for simpler code with less destructuring, particularly for single value outputs.
+export type IntentData<T extends IntentParams> =
+  S.Schema.Type<S.Struct<T>> extends { input: any } ? S.Schema.Type<S.Struct<T>>['input'] : any;
+
+export type IntentResultData<T extends IntentParams> =
+  S.Schema.Type<S.Struct<T>> extends { output: any } ? S.Schema.Type<S.Struct<T>>['output'] : any;
+
+export type IntentSchema<Tag extends string, Fields extends IntentParams> = S.TaggedClass<any, Tag, Fields>;
 
 /**
  * An intent is an abstract description of an operation to be performed.
  * Intents allow actions to be performed across plugins.
  */
-export type Intent = {
-  /**
-   * Plugin ID.
-   * If specified, the intent will be sent explicitly to the plugin.
-   * Otherwise, the intent will be sent to all plugins, in order and the first to resolve a non-null value will be used.
-   */
-  plugin?: string;
+export type Intent<
+  Tag extends string,
+  Fields extends IntentParams,
+  Schema extends IntentSchema<Tag, Fields> = IntentSchema<Tag, Fields>,
+> = {
+  _schema: Schema;
 
   /**
    * The action to perform.
    */
-  action: string;
-
-  /**
-   * Whether or not the intent is being undone.
-   */
-  undo?: boolean;
+  action: Tag;
 
   /**
    * Any data needed to perform the desired action.
    */
-  // TODO(burdon): Typed intents.
-  data?: IntentData;
-};
+  data: IntentData<Fields>;
 
-export type IntentResult = {
   /**
-   * The output of the action that was performed.
-   *
-   * If the intent is apart of a chain of intents, the data will be passed to the next intent.
+   * Plugin ID.
+   * If specified, the intent will be sent explicitly to the plugin.
+   * Otherwise, the intent will be sent to all plugins, in order and the first to resolve a non-null value will be used.
    */
-  data?: any;
+  plugin?: string;
 
   /**
-   * If provided, the action will be undoable.
+   * Whether or not the intent is being undone.
    */
-  undoable?: {
-    /**
-     * Message to display to the user when indicating that the action can be undone.
-     */
-    message: string;
-
-    /**
-     * Will be merged with the original intent data when firing the undo intent.
-     */
-    data?: IntentData;
-  };
+  undo?: boolean;
+};
 
-  /**
-   * An error that occurred while performing the action.
-   *
-   * If the intent is apart of a chain of intents and an error occurs, the chain will be aborted.
-   */
-  error?: Error;
+export type AnyIntent = Intent<any, any>;
 
-  /**
-   * Other intent chains to be triggered.
-   */
-  intents?: Intent[][];
+/**
+ * Chain of intents to be executed together.
+ * The result of each intent is merged into the next intent's input data.
+ */
+export type IntentChain<
```

Comment:
seems similar to `flow` from effect

Reply:
yeah, this could be used with flow if you wanted to create a reusable chain creator. Right now I'm just using it with pipe to create single chains.